sound/soc/sof/intel/hda.c

```diff
@@ -72,22 +72,38 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w, bool setup)
 	if (setup) {
 		ret = sof_widget_setup(sdev, swidget);
 		if (ret < 0) {
-			dev_err(sdev->dev, "error: setting up DAI widget %s\n", w->name);
+			dev_err(sdev->dev, "error: failed setting up DAI widget %s\n", w->name);
 			return ret;
 		}
 
 		/* send DAI_CONFIG IPC */
-		return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
+		ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
 					  &reply, sizeof(reply));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed setting DAI config for %s\n", w->name);
+			return ret;
+		}
+
+		sof_dai->configured = true;
+
+		return 0;
 	}
 
+	/*
+	 * nothing to do if hw_free() is called without restarting the stream after resume.
+	 */
+	if (!sof_dai->configured)
+		return 0;
+
 	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
 				  &reply, sizeof(reply));
 	if (ret < 0) {
-		dev_err(sdev->dev, "error: updating DAI config for %s\n", w->name);
+		dev_err(sdev->dev, "error: failed setting DAI config for %s\n", w->name);
```

Comment:
it's best to have unique message, ie. here this should be "failed setting DAI config for %s before widget_free", and the the other 'after widget setup'. It removes guess work on what might have happened